### PR TITLE
implement per-op config downstream changes

### PR DESCRIPTION
**Pull request by lucix-aws**
## Description of changes
* Add `ManagedCredentialsProvider` to handle share count tracking for defaulted providers in config.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/EcsCredentialsProvider.kt [hunk 1]
```diff
@@ -98,7 +98,9 @@ public class EcsCredentialsProvider internal constructor(
                 else -> CredentialsProviderException("Failed to get credentials from container metadata service", ex)
             }
         } finally {
-            client.close()
+            if (manageEngine) {
+                client.engine.close()
+            }
```
Comment (ianbotsf) [inline aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/EcsCredentialsProvider.kt:103]: **Question**: Why are we closing the engine after retrieving credentials? Might we have to retrieve credentials again in the future and won't that fail on a closed engine?

Same question for `ImdsClient`.

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/protocols/core/AwsHttpProtocolClientGenerator.kt [hunk 2]
```diff
@@ -37,6 +37,13 @@ open class AwsHttpProtocolClientGenerator(
         super.render(writer)
     }
 
+    override fun renderInit(writer: KotlinWriter) {
+        writer.withBlock("init {", "}") {
+            write("managedResources.#T(config.httpClientEngine)", RuntimeTypes.IO.addIfManaged)
+            write("managedResources.#T(config.credentialsProvider)", RuntimeTypes.IO.addIfManaged)
+        }
+    }
+
```
Comment (ianbotsf) [inline codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/protocols/core/AwsHttpProtocolClientGenerator.kt:46]: **Suggestion**: This feels like unnecessary duplication with `HttpProtocolClientGenerator.renderInit`. Consider a new function for subclasses to add additional things to the initializer or using sections.

**Review by ianbotsf — APPROVED** [review]

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/EcsCredentialsProvider.kt (hunk 1 above)
Comment (aajtodd) [reply aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/EcsCredentialsProvider.kt:103]: Ya this is wrong, previously we were passing `manageEngine = false` so the underlying engine wasn't closed when `SdkHttpClient` was closed. It seems like `SdkHttpClient` doesn't need `Closeable` anymore and doesn't need to be closed.

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/imds/ImdsClient.kt [hunk 3]
```diff
@@ -64,6 +64,7 @@ public class ImdsClient private constructor(builder: Builder) : InstanceMetadata
     private val platformProvider: PlatformProvider = builder.platformProvider
     private val sdkLogMode: SdkLogMode = builder.sdkLogMode
     private val httpClient: SdkHttpClient
+    private val manageEngine: Boolean = builder.engine == null
```
[the remark this answers is not in the document]
Comment (lucix-aws) [reply aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/imds/ImdsClient.kt:67]: Talked offline, and we want to strive for this in the future, but given how the HTTP client is spanned across several providers in the default chain (and then often passed a few levels down within the chain providers) there's a risk of getting the bookkeeping wrong somewhere in that environment, and it'll be safer to land/vet this separately.

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/internal/ManagedCredentialsProvider.kt [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+private class ManagedCredentialsProvider(
+    private val delegate: CredentialsProvider,
+) : SdkManagedCloseable(delegate), CredentialsProvider by delegate
+
+/**
+ * Wraps a [CredentialsProvider] for internal runtime management by the SDK.
+ */
+@InternalSdkApi
+public fun CredentialsProvider.manage(): CredentialsProvider =
```
Comment (aajtodd) [inline aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/internal/ManagedCredentialsProvider.kt:19]: **correctness**: I think this probably actually belongs in `smithy-kotlin` in same package as `CredentialsProvider`.

**Review by aajtodd — APPROVED** [review]

**Review by ianbotsf — APPROVED** [review]

**Pull request merged**